### ISSUE TITLE
maybe a memory leak in berry-spi

### ISSUE THREAD

**Issue by ghost** [Volantus/berry-spi#3]
First of all thank you for the work to program berry-spi.
It works as expected, only the available memory becomes less and less when it is running.
I use it on a raspberry pi 3 with the analog-digital-converter MCP3208.
I use the compiled version 0.4.1 php-7.0_berry_spi.so from https://github.com/Volantus/berry-spi/releases/tag/0.4.1 with PHP Version 7.0.33-0+deb9u1.
When the program is running, the available memory becomes less and less.
This can be tested with the following simple php-code in php-cli.
```
<?php
use Volantus\BerrySpi\RegularInterface;

$interface = new RegularInterface(1, 2000000, 0);
$interface->open();
$dummyData = [];
$dataToTransfer = [0x07, 0x40, 0x00]; //MCP3208 channel 5
`stty -icanon`;
stream_set_blocking(STDIN, 0);
echo "Press 'Q' to quit", PHP_EOL;
$nextLogTime = time() + 60; //1 minute
echo date('H:i:s'), PHP_EOL, `free`, PHP_EOL;
do {
  $dummyData = $interface->transfer($dataToTransfer);
  if (time() >= $nextLogTime) {
    echo date('H:i:s'), PHP_EOL, `free`, PHP_EOL;
    $nextLogTime = time() + 60; //1 minute
  }
} while (ord(fgetc(STDIN)) != 113);
$interface->close();
unset($interface);
echo "quit", PHP_EOL;
echo date('H:i:s'), PHP_EOL, `free`, PHP_EOL;
```
The program outputs the time and the output of free-command every minute.
With me the available memory becomes less and less.

Example outputs:
```
08:17:10
              total        used        free      shared  buff/cache   available
Mem:         994108      101000      760276       26804      132832      817028
Swap:       4202672           0     4202672

08:18:10
              total        used        free      shared  buff/cache   available
Mem:         994108      122532      738724       26804      132852      795476
Swap:       4202672           0     4202672

08:19:10
              total        used        free      shared  buff/cache   available
Mem:         994108      145160      716080       26804      132868      772852
Swap:       4202672           0     4202672

08:20:10
              total        used        free      shared  buff/cache   available
Mem:         994108      168496      692728       26804      132884      749516
Swap:       4202672           0     4202672

08:21:10
              total        used        free      shared  buff/cache   available
Mem:         994108      191368      669840       26804      132900      726644
Swap:       4202672           0     4202672

08:22:10
              total        used        free      shared  buff/cache   available
Mem:         994108      213724      647468       26804      132916      704304
Swap:       4202672           0     4202672

08:23:10
              total        used        free      shared  buff/cache   available
Mem:         994108      237068      624092       26804      132948      680948
Swap:       4202672           0     4202672

quit
08:23:29
              total        used        free      shared  buff/cache   available
Mem:         994108      244516      616644       26804      132948      673508
Swap:       4202672           0     4202672
```
After the program has been completely stopped with the "q" key, the memory is fully released again.
The free-command shows that the available memory is again approximately as before the start of the program.
```
              total        used        free      shared  buff/cache   available
Mem:         994108       96668      764488       26800      132952      821360
Swap:       4202672           0     4202672
```
If the analog-digital-converter queries run for several hours/days ..., swap is used until there is no more free memory and the raspberry crashes.

Is the memory leak in berry-spi or am i doing something wrong?

Many thanks for the help!
Joerg

**Comment by marius-meissner** [comment]
Thank you for the detailed report!
The mentioned PHP code is correct, so you detected a memory leak.

Seems there is a destructor for the [transmission buffer](https://github.com/Volantus/berry-spi/blob/76262b1dc2586dd0a7fa8833176ce5f56ceb8212/src/TransmitBuffer.hpp#L9) (char pointer) missing.
I will create a branch addressing this issue tomorrow and will keep you up2date.

**Comment by marius-meissner** [comment]
@jborch01 The missing constructor has been added in release [0.5.0](https://github.com/Volantus/berry-spi/releases/tag/0.5.0). Please test if the memory leak still exists with this version.
Thank you!

**Comment by ghost** [comment]
Hello SkydiveMarius,

but can you please add the PHP7.0 support again.
The standard PHP in raspberry's Debian Stretch is PHP 7.0 (for now 7.0.33).

Thank you!

**Comment by marius-meissner** [comment]
@jborch01 Unfortunately I do not have an environment for PHP 7.0.
But the binary is dynamically linked and should work also for PHP 7.0 as long as you have PHP-CPP compiled with the correct PHP version.

As an alternative you may compile the extension [on you own](https://github.com/Volantus/berry-spi#compile-from-source).

**Comment by ghost** [comment]
Thank you Marius,

everything works now, the memory leak is closed.
I have compiled the extension for PHP7.0 on my own.
It did not work right away with the Makefile.
I still had to change the paths in the Makefile:
```
from 
COMPILER_FLAGS = -Wall -I/usr/local/src/php -I/usr/local/src/php/main -I/usr/local/src/php/Zend -I/usr/local/src/php/TSRM -c -O2 -std=c++11 -fpic -o

to
COMPILER_FLAGS = -Wall -I/usr/include -I/usr/include/php -I/usr/include/php/20151012 -I/usr/include/php/20151012/Zend -I/usr/include/php/20151012/TSRM -c -O2 -std=c++11 -fpic -o
```
I'm not very familiar with compiling programs, but I think that's right because compiling has worked with this changes.
Is it possible and/or desired that I upload the for php7.0 compiled extension php-7.0_berry-spi.so here? Maybe that's helpful for other users.

Thank you for programming berry-spi!
The issue can be closed.

Joerg

**Comment by marius-meissner** [comment]
Great to hear that the the issue was fixed and works with PHP 7.0.

> Is it possible and/or desired that I upload the for php7.0 compiled extension php-7.0_berry-spi.so here? 

That would be helpful. Please upload the binary to Github (e.g. create a tmp. GIT repository) and I will add the binary to the release then.

**Comment by ghost** [comment]
> Great to hear that the the issue was fixed and works with PHP 7.0.
> 
> > Is it possible and/or desired that I upload the for php7.0 compiled extension php-7.0_berry-spi.so here?
> 
> That would be helpful. Please upload the binary to Github (e.g. create a tmp. GIT repository) and I will add the binary to the release then.

@SkydiveMarius 
You find the php-7.0_berry-spi.so [here](https://github.com/jborch01/temp_for_php_spi) .
If you have downloaded the file, I delete my repository.
Please add the binary to your release of berry-spi.

Thank you!

**Comment by marius-meissner** [comment]
Thank you, binary has been added to the release.
